module/zfs/dmu_objset.c

```diff
@@ -1639,20 +1640,67 @@ sync_dnodes_task(void *arg)
 
 	multilist_sublist_unlock(ms);
 
+	zio_nowait(sda->sda_cio);
+
 	kmem_free(sda, sizeof (*sda));
 }
 
+typedef struct sync_objset_arg {
+	objset_t *soa_os;
+	dmu_tx_t *soa_tx;
+	zio_t *soa_zio;
+	taskq_ent_t soa_tq_ent;
+} sync_objset_arg_t;
+
+static void
+sync_dnodes_finish_task(void *arg)
+{
+	sync_objset_arg_t *soa = arg;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+
+	/* Free intent log blocks up to this tx. (May block.) */
+	zil_sync(os->os_zil, tx);
+	os->os_phys->os_zil_header = os->os_zil_header;
+
+	zio_nowait(soa->soa_zio);
+	kmem_free(soa, sizeof (*soa));
+}
 
-/* called from dsl */
 void
-dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
+dmu_objset_sync_sublists_done(zio_t *zio)
+{
+	sync_objset_arg_t *soa = zio->io_private;
+	objset_t *os = soa->soa_os;
+	dmu_tx_t *tx = soa->soa_tx;
+	int txgoff = tx->tx_txg & TXG_MASK;
+	dbuf_dirty_record_t *dr;
+
+	list_t *list = &DMU_META_DNODE(os)->dn_dirty_records[txgoff];
+	while ((dr = list_remove_head(list)) != NULL) {
+		ASSERT0(dr->dr_dbuf->db_level);
+		zio_nowait(dr->dr_zio);
+	}
+
+	/* Enable dnode backfill if enough objects have been freed. */
+	if (os->os_freed_dnodes >= dmu_rescan_dnode_threshold) {
+		os->os_rescan_dnodes = B_TRUE;
+		os->os_freed_dnodes = 0;
+	}
+
+	/* sync_dnodes_finsh_task calls zil_sync on our behalf. */
+	taskq_dispatch_ent(dmu_objset_pool(os)->dp_sync_taskq,
+	    sync_dnodes_finish_task, soa, TQ_FRONT, &soa->soa_tq_ent);
+}
+
+/* Nonblocking objset sync. Called from dsl. */
+zio_t *
```

Comment:
Making dmu_objset_sync() not block on zio completions is useful in the case such as a zvol. When a dataset has fewer objsets than syncthreads, then the pool sync can perform more syncing operations in parallel, rather than wait for each objset to complete before starting on the next dataset.